### PR TITLE
Cache `study.directions` to reduce the number of `get_study_directions()` calls.

### PR DESCRIPTION
**Pull request by c-bata**
## Motivation

As discussed at https://github.com/optuna/optuna/pull/2023#issuecomment-1303069347, `study.directions` property method is called many times during the optimization. `study.directions` is set when creating a study via the `optuna.create_study(..., directions=...)` function and never be modified, so that we can obviously use `cached_property` and reduce the calls of `storage.get_study_directions()`.

## Description of the changes

~~Use `functools.cached_property` instead of `property`.~~

`functools.cached_property` seemed to be introduced in Python 3.8. So I pushed [1e145ca](https://github.com/optuna/optuna/pull/4146/commits/1e145caeb825ac62eff8013650202920fa185739).

### CODE REVIEW

**Comment by not522** [comment]
I'm not confident, but what about using `lru_cache`?

**Comment by c-bata** [comment]
Thank you for the suggestion!
It sounds a bit not intuitive since this is a property method and there are no arguments. What do you think?

**Comment by not522** [comment]
Hmm, I agree `lru_cache` is not intuitive.
On second thought, how about making `directions` the `Study`'s attribute? We can set it in `__init__`.

**Comment by c-bata** [comment]
Sounds reasonable! One possible concern is that users may try to rewrite `study.directions` like `study.directions = [...]`. Do you think it's acceptable?

If it's not acceptable, we can set `study._directions` in `__init__` and provide users via the property method.

**Comment by not522** [comment]
> If it's not acceptable, perhaps we can set `study._directions` in `__init__` and provide users via the property method.

I favor this plan.

File: optuna/study/study.py [hunk 1]
```diff
@@ -82,6 +82,7 @@ def __init__(
         study_id = storage.get_study_id_from_name(study_name)
         self._study_id = study_id
         self._storage = storage
+        self._directions: Optional[List[StudyDirection]] = None
```
Comment (c-bata) [inline optuna/study/study.py:85]:
```suggestion
        self._directions = storage.get_study_directions(study_id)
```

File: optuna/study/study.py [hunk 2]
```diff
@@ -202,7 +203,9 @@ def directions(self) -> List[StudyDirection]:
             A list of :class:`~optuna.study.StudyDirection` objects.
         """
 
-        return self._storage.get_study_directions(self._study_id)
+        if self._directions is None:
+            self._directions = self._storage.get_study_directions(self._study_id)
+        return self._directions
```
Comment (c-bata) [inline optuna/study/study.py:208]:
```suggestion
        return self._directions
```

**Comment by c-bata** [comment]
PTAL.

**Review by not522 — APPROVED** [review]
LGTM

Note: we can assume users don't directly create `Study`. The document says

> Note that the direct use of this constructor is not recommended.

**Comment by toshihikoyanase** [comment]
@contramundum53 Could you review this PR, please?

File: optuna/study/study.py [hunk 3]
```diff
@@ -1169,10 +1170,9 @@ def objective(trial):
         sampler = samplers.NSGAIISampler()
 
     study_name = storage.get_study_name_from_id(study_id)
+    storage.set_study_directions(study_id, direction_objects)
```
Comment (contramundum53) [inline optuna/study/study.py:1173]: Almost LGTM, but I'm not confident about the consequence of moving this line before creating `Study`. Could you explain the motivation behind this?

**Comment by contramundum53** [comment]
There seems to be a slight risk of creating an inconsistent state when more than two processes create the study at the same time. I don't think we need to address that, but maybe we should mention that in the document?

File: optuna/study/study.py (hunk 3 above)
Comment (c-bata) [reply optuna/study/study.py:1173]: In order to apply [this change](https://github.com/optuna/optuna/pull/4146#issuecomment-1308045371), we must ensure that directions must be set before instantiating `Study` object. Is my answer what you need?

**Comment by c-bata** [comment]
> There seems to be a slight risk of creating an inconsistent state when more than two processes create the study at the same time with different directions. I don't think we need to address that, but maybe we should mention that in the document?

Could you tell me more details on how to create inconsistent studies with more than two processes?

File: optuna/study/study.py (hunk 3 above)
Comment (contramundum53) [reply optuna/study/study.py:1173]: I see, thanks!

**Comment by contramundum53** [comment]
A race condition is hit when
1. Process A creates a study
2. Process B creates the same study
3. Process A sets a direction in the storage
4. Process B sets a different direction in the storage.
I think the consequence is that two processes continues optimization for different directions, which is not what we expect. (We expect an error in this case.)

**Comment by contramundum53** [comment]
Well, is there currently any protection against inconsistent directions at all?
If not, maybe we could add that in the future.

**Comment by contramundum53** [comment]
It seems that `Storage` has protection against inconsistent directions in it.

**Review by contramundum53 — APPROVED** [review]
LGTM! Thanks for the PR!

**Pull request merged**